Port blk_rw_mt test to Windows

ChandKV opened the pull request:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/pmem/nvml/979)
<!-- Reviewable:end -->

plebioda commented:
Reviewed 19 of 19 files at r1.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[a discussion](https://reviewable.io:443/reviews/pmem/nvml/979#-KMTKhh9ul1adM0OAdFP:-KMTKhhAmCF2J19kKHaT:60559717) (no related file):*
Please change the *.ps1 to *.PS1.

---

*[src/common/pthread_windows.c, line 384 \[r1\]](https://reviewable.io:443/reviews/pmem/nvml/979#-KMTBMmW093oJ5ttPoOH:-KMTBMmXAyu3sZA-A-Br:-571854398) ([raw file](https://github.com/pmem/nvml/blob/547e8f63fd769ff4e07ec41afd19244ddb3b3325/src/common/pthread_windows.c#L384)):*
> ```C
> {
> 	pthread_self_index = TlsAlloc();
> 	if (pthread_self_index == TLS_OUT_OF_INDEXES) abort();
> ```

Please move abort() to new line.

---

*[src/common/pthread_windows.c, line 411 \[r1\]](https://reviewable.io:443/reviews/pmem/nvml/979#-KMTKK42KqYmEmo67VmH:-KMTKK42KqYmEmo67VmI:-1891566588) ([raw file](https://github.com/pmem/nvml/blob/547e8f63fd769ff4e07ec41afd19244ddb3b3325/src/common/pthread_windows.c#L411)):*
> ```C
> 	void *(*start_routine)(void *), void *arg)
> {
> 	pthread_t thread_info;
> ```

Can't we just fill the thread structure passed by user ?

---

*[src/common/pthread_windows.c, line 413 \[r1\]](https://reviewable.io:443/reviews/pmem/nvml/979#-KMTHATyx-tJiR1syzQM:-KMTHATyx-tJiR1syzQN:-670185357) ([raw file](https://github.com/pmem/nvml/blob/547e8f63fd769ff4e07ec41afd19244ddb3b3325/src/common/pthread_windows.c#L413)):*
> ```C
> 	pthread_t thread_info;
> 
> 	assert(attr == NULL);
> ```

please use ASSERT()  or ASSERTeq(attr, NULL)

---

*[src/test/blk_rw_mt/TEST0.ps1, line 59 \[r1\]](https://reviewable.io:443/reviews/pmem/nvml/979#-KMTJF402cwYuodnfWiW:-KMTJF402cwYuodnfWiX:-237562878) ([raw file](https://github.com/pmem/nvml/blob/547e8f63fd769ff4e07ec41afd19244ddb3b3325/src/test/blk_rw_mt/TEST0.ps1#L59)):*
> ```PowerShell
> expect_normal_exit ..\..\x64\debug\blk_rw_mt$EXESUFFIX 4096 $DIR\testfile1 123 5 80
> 
> # check_pool $DIR\testfile1
> ```

I would suggest to leave XXX + comment why this is commented. It will be much easier to find all lines which should be uncommented after implementing required functionality.
The same issue in other *.ps1 files.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/979)*
<!-- Sent from Reviewable.io -->

tomaszkapela commented:
Review status: all files reviewed at latest revision, 10 unresolved discussions.

---

*[src/common/pthread_windows.c, line 415 \[r1\]](https://reviewable.io:443/reviews/pmem/nvml/979#-KMTKr-74H-9mQA4n2sF:-KMTKr-8gQ9VYluV6KCM:-17172458) ([raw file](https://github.com/pmem/nvml/blob/547e8f63fd769ff4e07ec41afd19244ddb3b3325/src/common/pthread_windows.c#L415)):*
> ```C
> 	assert(attr == NULL);
> 
> 	if ((thread_info = malloc(sizeof(pthread_info))) == NULL)
> ```

This evaluates `pthread_t var = malloc();` and to me this looks really strange. The typedef hides the pointer type. Besides that, you declare a pthread_t and malloc with sizeof(pthread_info) which is also confusing.

---

*[src/common/pthread_windows.c, line 425 \[r1\]](https://reviewable.io:443/reviews/pmem/nvml/979#-KMTLWFXeszVOkTQYSjp:-KMTLWFXeszVOkTQYSjq:141082255) ([raw file](https://github.com/pmem/nvml/blob/547e8f63fd769ff4e07ec41afd19244ddb3b3325/src/common/pthread_windows.c#L425)):*
> ```C
> 		NULL);
> 	if (thread_info->threadHandle == 0)
> 		return errno;
> ```

return without free

---

*[src/common/pthread_windows.c, line 428 \[r1\]](https://reviewable.io:443/reviews/pmem/nvml/979#-KMTLYFXCv4RpKRdobQ1:-KMTLYFXCv4RpKRdobQ2:141082255) ([raw file](https://github.com/pmem/nvml/blob/547e8f63fd769ff4e07ec41afd19244ddb3b3325/src/common/pthread_windows.c#L428)):*
> ```C
> 
> 	if (ResumeThread(thread_info->threadHandle) == -1)
> 		return EAGAIN;
> ```

return without free

---

*[src/common/pthread_windows.c, line 444 \[r1\]](https://reviewable.io:443/reviews/pmem/nvml/979#-KMTLb2cOlrrss0Q1v4O:-KMTLb2cOlrrss0Q1v4P:1490249001) ([raw file](https://github.com/pmem/nvml/blob/547e8f63fd769ff4e07ec41afd19244ddb3b3325/src/common/pthread_windows.c#L444)):*
> ```C
> 		*result = thread->result;
> 
> 	free(thread);
> ```

I know that maybe this is not the case in our tests, but you could let the threads run loose and never join them, thus never freeing this struct.

---

*[src/windows/include/pthread.h, line 159 \[r1\]](https://reviewable.io:443/reviews/pmem/nvml/979#-KMTLxx2rE1onzyAcI0V:-KMTLxx2rE1onzyAcI0W:-1804062604) ([raw file](https://github.com/pmem/nvml/blob/547e8f63fd769ff4e07ec41afd19244ddb3b3325/src/windows/include/pthread.h#L159)):*
> ```objc
> 	void *(*startRoutine)(void *);
> 	void *result;
> } pthread_info, * pthread_t;
> ```

I would suggest removing the pthread_info altogether and change the pthread_t from pointer to a normal type.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/979)*
<!-- Sent from Reviewable.io -->

lplewa commented:
Review status: all files reviewed at latest revision, 11 unresolved discussions.

---

*[src/windows/include/platform.h, line 251 \[r1\]](https://reviewable.io:443/reviews/pmem/nvml/979#-KMTN-n4LHa83nTV-QEe:-KMTN-n4LHa83nTV-QEf:-538969661) ([raw file](https://github.com/pmem/nvml/blob/547e8f63fd769ff4e07ec41afd19244ddb3b3325/src/windows/include/platform.h#L251)):*
> ```objc
> #ifdef _MT
> #define rand_r rand_s
> #endif // _MT
> ```

Please don't use c++ // comments 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/979)*
<!-- Sent from Reviewable.io -->

krzycz commented:
Reviewed 15 of 19 files at r1.
Review status: all files reviewed at latest revision, 15 unresolved discussions.

---

*[src/common/pthread_windows.c, line 369 \[r1\]](https://reviewable.io:443/reviews/pmem/nvml/979#-KMTN1Z6C-Me1O0sb-YR:-KMTN1Z6C-Me1O0sb-YS:-41419039) ([raw file](https://github.com/pmem/nvml/blob/547e8f63fd769ff4e07ec41afd19244ddb3b3325/src/common/pthread_windows.c#L369)):*
> ```C
> 
> /* threading */
> pthread_once_t pthread_self_index_initialized;
> ```

In our coding style, variables with global scope should have a name with first letter capitalized: Pthread_self_index_initialized.

---

*[src/test/blk_rw_mt/blk_rw_mt.c, line 90 \[r1\]](https://reviewable.io:443/reviews/pmem/nvml/979#-KMTOHJ1yGb8EQFaEIt7:-KMTOHJ1yGb8EQFaEIt8:-1279651643) ([raw file](https://github.com/pmem/nvml/blob/547e8f63fd769ff4e07ec41afd19244ddb3b3325/src/test/blk_rw_mt/blk_rw_mt.c#L90)):*
> ```C
> 	int ord = 1;
> 
> 	for (unsigned int i = 0; i < Nops; i++) {
> ```

1) Any reason for "unsigned" here?
2) Please, just "unsigned", not "unsigned int".

---

*[src/test/blk_rw_mt/blk_rw_mt.c, line 140 \[r1\]](https://reviewable.io:443/reviews/pmem/nvml/979#-KMTPe86Ptm_VGJsgW_V:-KMTPe86Ptm_VGJsgW_W:211775766) ([raw file](https://github.com/pmem/nvml/blob/547e8f63fd769ff4e07ec41afd19244ddb3b3325/src/test/blk_rw_mt/blk_rw_mt.c#L140)):*
> ```C
> 	/* kick off nthread threads */
> 	for (unsigned int i = 0; i < Nthread; i++)
> 		PTHREAD_CREATE(&threads[i], NULL, worker, (void *)(intptr_t)i);
> ```

So, we cast *unsigned* to *intptr_t*, then to *void **, then to *intptr_t* again, and eventually to *long*. :-)
Perhaps, it would be better to use *long long* (or *intptr_t*) for *mytid*, to make it pointer size. This would eliminate need for one of these casts, I believe.

---

*[src/windows/include/platform.h, line 251 \[r1\]](https://reviewable.io:443/reviews/pmem/nvml/979#-KMTJu_OwTi7uwRi6z6e:-KMTJu_OwTi7uwRi6z6f:-1259116695) ([raw file](https://github.com/pmem/nvml/blob/547e8f63fd769ff4e07ec41afd19244ddb3b3325/src/windows/include/platform.h#L251)):*
> ```objc
> #ifdef _MT
> #define rand_r rand_s
> #endif // _MT
> ```

Please, do not use C++-style comments.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/979)*
<!-- Sent from Reviewable.io -->

ChandKV commented:
Review status: 11 of 19 files reviewed at latest revision, 15 unresolved discussions, some commit checks failed.

---

*[a discussion](https://reviewable.io:443/reviews/pmem/nvml/979#-KMTKhh9ul1adM0OAdFP:-KMVGq9rjmpq3T6deIpa:676635175) (no related file):*
<details><summary><i>Previously, plebioda wrote…</i></summary>
> Please change the *.ps1 to *.PS1.
</details>
Done

---

*[src/common/pthread_windows.c, line 369 \[r1\]](https://reviewable.io:443/reviews/pmem/nvml/979#-KMTN1Z6C-Me1O0sb-YR:-KMV1Pevo3F4BjRFdK4T:-499146009) ([raw file](https://github.com/pmem/nvml/blob/547e8f63fd769ff4e07ec41afd19244ddb3b3325/src/common/pthread_windows.c#L369)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary>
> In our coding style, variables with global scope should have a name with first letter capitalized: Pthread_self_index_initialized.
</details>
Done.

---

*[src/common/pthread_windows.c, line 384 \[r1\]](https://reviewable.io:443/reviews/pmem/nvml/979#-KMTBMmW093oJ5ttPoOH:-KMVGdNDHqsPM2ICUgJV:-499146009) ([raw file](https://github.com/pmem/nvml/blob/547e8f63fd769ff4e07ec41afd19244ddb3b3325/src/common/pthread_windows.c#L384)):*
<details><summary><i>Previously, plebioda wrote…</i></summary>
> Please move abort() to new line.
</details>
Done.

---

*[src/common/pthread_windows.c, line 411 \[r1\]](https://reviewable.io:443/reviews/pmem/nvml/979#-KMTKK42KqYmEmo67VmH:-KMUtUM9synlXkZ5DSTd:-193529364) ([raw file](https://github.com/pmem/nvml/blob/547e8f63fd769ff4e07ec41afd19244ddb3b3325/src/common/pthread_windows.c#L411)):*
<details><summary><i>Previously, plebioda wrote…</i></summary>
> Can't we just fill the thread structure passed by user ?
</details>
We can but then in failure path *thread will have a residue. We may choose to stamp a NULL in that case. Here the choice is to leave the original value in case of failure.

---

*[src/common/pthread_windows.c, line 413 \[r1\]](https://reviewable.io:443/reviews/pmem/nvml/979#-KMTHATyx-tJiR1syzQM:-KMVGhWcEShpeCpWFM08:-499146009) ([raw file](https://github.com/pmem/nvml/blob/547e8f63fd769ff4e07ec41afd19244ddb3b3325/src/common/pthread_windows.c#L413)):*
<details><summary><i>Previously, plebioda wrote…</i></summary>
> please use ASSERT()  or ASSERTeq(attr, NULL)
</details>
Done.

---

*[src/common/pthread_windows.c, line 415 \[r1\]](https://reviewable.io:443/reviews/pmem/nvml/979#-KMTKr-74H-9mQA4n2sF:-KMV2KGjUQDQ-ShBGYWO:-1489750148) ([raw file](https://github.com/pmem/nvml/blob/547e8f63fd769ff4e07ec41afd19244ddb3b3325/src/common/pthread_windows.c#L415)):*
<details><summary><i>Previously, tomaszkapela wrote…</i></summary>
> This evaluates `pthread_t var = malloc();` and to me this looks really strange. The typedef hides the pointer type. Besides that, you declare a pthread_t and malloc with sizeof(pthread_info) which is also confusing.
</details>
I believe the below is better for readability, let me know your thoughts.

	pthread_info *thread_info;
	if ((thread_info = malloc(sizeof(pthread_info))) == NULL)
...
	*thread = (pthread_t)thread_info;

---

*[src/common/pthread_windows.c, line 425 \[r1\]](https://reviewable.io:443/reviews/pmem/nvml/979#-KMTLWFXeszVOkTQYSjp:-KMV2hPb4_xew_FWEC6S:-460479470) ([raw file](https://github.com/pmem/nvml/blob/547e8f63fd769ff4e07ec41afd19244ddb3b3325/src/common/pthread_windows.c#L425)):*
<details><summary><i>Previously, tomaszkapela wrote…</i></summary>
> return without free
</details>
Thank you, fixed.

---

*[src/common/pthread_windows.c, line 428 \[r1\]](https://reviewable.io:443/reviews/pmem/nvml/979#-KMTLYFXCv4RpKRdobQ1:-KMV2nkhLXZoZ5k78Dzy:-460479470) ([raw file](https://github.com/pmem/nvml/blob/547e8f63fd769ff4e07ec41afd19244ddb3b3325/src/common/pthread_windows.c#L428)):*
<details><summary><i>Previously, tomaszkapela wrote…</i></summary>
> return without free
</details>
Thank you, fixed.

---

*[src/common/pthread_windows.c, line 444 \[r1\]](https://reviewable.io:443/reviews/pmem/nvml/979#-KMTLb2cOlrrss0Q1v4O:-KMV2wO0fKkNZgANifQz:1252811937) ([raw file](https://github.com/pmem/nvml/blob/547e8f63fd769ff4e07ec41afd19244ddb3b3325/src/common/pthread_windows.c#L444)):*
<details><summary><i>Previously, tomaszkapela wrote…</i></summary>
> I know that maybe this is not the case in our tests, but you could let the threads run loose and never join them, thus never freeing this struct.
</details>
That's by design, right? If someone creates a joinable thread and doesn't join we are bound to leak resources right (zombie threads). In the minimum we need to store the thread's exit status somewhere so that we can hand that over to a thread that calls join. Also we need to reserve the index or thread-id, so that a later thread won't take its place.

If someone is not interested in waiting for a thread, they should create a DETACHED thread. And because our tests are not doing that, yet, our pthread implementation doesn't implement the PTHREAD_CREATE_DETACHED attribute (any attribute for that matter).

---

*[src/test/blk_rw_mt/blk_rw_mt.c, line 90 \[r1\]](https://reviewable.io:443/reviews/pmem/nvml/979#-KMTOHJ1yGb8EQFaEIt7:-KMV5CLvq4xvExPi478a:-1873956208) ([raw file](https://github.com/pmem/nvml/blob/547e8f63fd769ff4e07ec41afd19244ddb3b3325/src/test/blk_rw_mt/blk_rw_mt.c#L90)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary>
> 1) Any reason for "unsigned" here?
> 2) Please, just "unsigned", not "unsigned int".
</details>
"Nops" is unsigned and compiler is not very happy comparing signed with unsigned. Changed "unsigned int" to "unsigned".

---

*[src/test/blk_rw_mt/blk_rw_mt.c, line 140 \[r1\]](https://reviewable.io:443/reviews/pmem/nvml/979#-KMTPe86Ptm_VGJsgW_V:-KMV6QFt8uFF-aJEe4nI:-1779603697) ([raw file](https://github.com/pmem/nvml/blob/547e8f63fd769ff4e07ec41afd19244ddb3b3325/src/test/blk_rw_mt/blk_rw_mt.c#L140)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary>
> So, we cast *unsigned* to *intptr_t*, then to *void **, then to *intptr_t* again, and eventually to *long*. :-)
> Perhaps, it would be better to use *long long* (or *intptr_t*) for *mytid*, to make it pointer size. This would eliminate need for one of these casts, I believe.
</details>
We can use "intptr_t" as the type for Nthread and the index variable. Hope that's desirable as well. However, changing mytid to intptr_t would just defer the cast till next line where "myseed" is unsinged and we have to cast "mytid" to long or so. And "myseed" needs to be unsigned as rand_r expects it to.

---

*[src/windows/include/platform.h, line 251 \[r1\]](https://reviewable.io:443/reviews/pmem/nvml/979#-KMTJu_OwTi7uwRi6z6e:-KMVGkXvH03cVrPSMPy-:-499146009) ([raw file](https://github.com/pmem/nvml/blob/547e8f63fd769ff4e07ec41afd19244ddb3b3325/src/windows/include/platform.h#L251)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary>
> Please, do not use C++-style comments.
</details>
Done.

---

*[src/windows/include/platform.h, line 251 \[r1\]](https://reviewable.io:443/reviews/pmem/nvml/979#-KMTN-n4LHa83nTV-QEe:-KMVGlyR5fcdK3TxG91h:-499146009) ([raw file](https://github.com/pmem/nvml/blob/547e8f63fd769ff4e07ec41afd19244ddb3b3325/src/windows/include/platform.h#L251)):*
<details><summary><i>Previously, lplewa (Łukasz Plewa) wrote…</i></summary>
> Please don't use c++ // comments 
</details>
Done.

---

*[src/windows/include/pthread.h, line 159 \[r1\]](https://reviewable.io:443/reviews/pmem/nvml/979#-KMTLxx2rE1onzyAcI0V:-KMV8Kod6uInip5srkn5:611785195) ([raw file](https://github.com/pmem/nvml/blob/547e8f63fd769ff4e07ec41afd19244ddb3b3325/src/windows/include/pthread.h#L159)):*
<details><summary><i>Previously, tomaszkapela wrote…</i></summary>
> I would suggest removing the pthread_info altogether and change the pthread_t from pointer to a normal type.
</details>
Then the caller of pthread_create() has to allocate memory for pthread_t. That is the calling code should be different for LINUX and Windows. Isn't the new way acceptable?

	pthread_info *thread_info;
	if ((thread_info = malloc(sizeof(pthread_info))) == NULL)
...
	*thread = (pthread_t)thread_info;

---

*[src/test/blk_rw_mt/TEST0.ps1, line 59 \[r1\]](https://reviewable.io:443/reviews/pmem/nvml/979#-KMTJF402cwYuodnfWiW:-KMV7XndxzDWqtt8gfJh:718122202) ([raw file](https://github.com/pmem/nvml/blob/547e8f63fd769ff4e07ec41afd19244ddb3b3325/src/test/blk_rw_mt/TEST0.ps1#L59)):*
<details><summary><i>Previously, plebioda wrote…</i></summary>
> I would suggest to leave XXX + comment why this is commented. It will be much easier to find all lines which should be uncommented after implementing required functionality.
> The same issue in other *.ps1 files.
</details>
I see XXX is used as TODO, interesting. Fixed.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/979)*
<!-- Sent from Reviewable.io -->

krzycz commented:
Reviewed 10 of 11 files at r2.
Review status: 18 of 19 files reviewed at latest revision, 12 unresolved discussions.

---

*[src/test/blk_rw_mt/blk_rw_mt.c, line 140 \[r1\]](https://reviewable.io:443/reviews/pmem/nvml/979#-KMTPe86Ptm_VGJsgW_V:-KMZOJKGz5mMAENOhk0z:2094797307) ([raw file](https://github.com/pmem/nvml/blob/547e8f63fd769ff4e07ec41afd19244ddb3b3325/src/test/blk_rw_mt/blk_rw_mt.c#L140)):*
<details><summary><i>Previously, ChandKV (Chandra Kumar) wrote…</i></summary>
> We can use "intptr_t" as the type for Nthread and the index variable. Hope that's desirable as well. However, changing mytid to intptr_t would just defer the cast till next line where "myseed" is unsinged and we have to cast "mytid" to long or so. And "myseed" needs to be unsigned as rand_r expects it to.
</details>
Ha! So now it looks like the first version was the best. ;-)

---

*[src/windows/include/pthread.h, line 159 \[r1\]](https://reviewable.io:443/reviews/pmem/nvml/979#-KMTLxx2rE1onzyAcI0V:-KM_5SRccjbhxLnX58jD:1338414385) ([raw file](https://github.com/pmem/nvml/blob/547e8f63fd769ff4e07ec41afd19244ddb3b3325/src/windows/include/pthread.h#L159)):*
<details><summary><i>Previously, ChandKV (Chandra Kumar) wrote…</i></summary>
> Then the caller of pthread_create() has to allocate memory for pthread_t. That is the calling code should be different for LINUX and Windows. Isn't the new way acceptable?
> 
> 	pthread_info *thread_info;
> 	if ((thread_info = malloc(sizeof(pthread_info))) == NULL)
> ...
> 	*thread = (pthread_t)thread_info;
</details>
I don't think it should be different. 
I agree with @tomaszkapela, that pthread_create() should not allocate memory for pthread_t. This is the caller duty, I think.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/979)*
<!-- Sent from Reviewable.io -->

ChandKV commented:
Review status: 18 of 19 files reviewed at latest revision, 12 unresolved discussions.

---

*[src/test/blk_rw_mt/blk_rw_mt.c, line 140 \[r1\]](https://reviewable.io:443/reviews/pmem/nvml/979#-KMTPe86Ptm_VGJsgW_V:-KM_FWfqWLNI0HYLWmjH:24816222) ([raw file](https://github.com/pmem/nvml/blob/547e8f63fd769ff4e07ec41afd19244ddb3b3325/src/test/blk_rw_mt/blk_rw_mt.c#L140)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary>
> Ha! So now it looks like the first version was the best. ;-)
</details>
Yes, it was lil more code but more explicit. This version silently converts intptr_t to long. I take this as you liked the first version. Thank you, will revert this.

---

*[src/windows/include/pthread.h, line 159 \[r1\]](https://reviewable.io:443/reviews/pmem/nvml/979#-KMTLxx2rE1onzyAcI0V:-KM_Gpmib-HNNS6PVvKy:-355886856) ([raw file](https://github.com/pmem/nvml/blob/547e8f63fd769ff4e07ec41afd19244ddb3b3325/src/windows/include/pthread.h#L159)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary>
> I don't think it should be different. 
> I agree with @tomaszkapela, that pthread_create() should not allocate memory for pthread_t. This is the caller duty, I think.
</details>

pthread_t thread;
pthread_create(&thread, NULL, start_bus, NULL);
pthread_create(&thread, NULL, start_car, NULL);
pthread_exit(NULL);

The above piece of code should successfully call spawn one thread to execute start_bus() and another to execute start_car(). And because the main thread doesn't call pthread_join() we could leak resources but there shouldn't be any uncertainty in what the threads execute.

If we let the callers allocate memory for pthread_t, the above piece of code could run start_car twice, or crash.

Also if we had to implement PTHREAD_CREATE_DETACHED tomorrow, above sequence (with PTHREAD_CREATE_DETACHED attribute) should be considered more common.

Let me know your suggestion.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/979)*
<!-- Sent from Reviewable.io -->

ChandKV commented:
Review status: 11 of 19 files reviewed at latest revision, 12 unresolved discussions.

---

*[src/common/pthread_windows.c, line 411 \[r1\]](https://reviewable.io:443/reviews/pmem/nvml/979#-KMTKK42KqYmEmo67VmH:-KMon_MreYOYOsNxTNOE:-1352226541) ([raw file](https://github.com/pmem/nvml/blob/547e8f63fd769ff4e07ec41afd19244ddb3b3325/src/common/pthread_windows.c#L411)):*
<details><summary><i>Previously, ChandKV (Chandra Kumar) wrote…</i></summary>
> We can but then in failure path *thread will have a residue. We may choose to stamp a NULL in that case. Here the choice is to leave the original value in case of failure.
</details>
I assume this can be considered resolved. Please let me know otherwise.

---

*[src/common/pthread_windows.c, line 415 \[r1\]](https://reviewable.io:443/reviews/pmem/nvml/979#-KMTKr-74H-9mQA4n2sF:-KMonjQ67siAo63llm6Y:1527161111) ([raw file](https://github.com/pmem/nvml/blob/547e8f63fd769ff4e07ec41afd19244ddb3b3325/src/common/pthread_windows.c#L415)):*
<details><summary><i>Previously, ChandKV (Chandra Kumar) wrote…</i></summary>
> I believe the below is better for readability, let me know your thoughts.
> 
> 	pthread_info *thread_info;
> 	if ((thread_info = malloc(sizeof(pthread_info))) == NULL)
> ...
> 	*thread = (pthread_t)thread_info;
</details>
I think this is partly tied to the discussion where we are evaluating moving the allocation to the caller. So I believe this can be marked resolved. Please let me know otherwise.

---

*[src/common/pthread_windows.c, line 444 \[r1\]](https://reviewable.io:443/reviews/pmem/nvml/979#-KMTLb2cOlrrss0Q1v4O:-KMoo597yks8iH72VtXL:54417514) ([raw file](https://github.com/pmem/nvml/blob/547e8f63fd769ff4e07ec41afd19244ddb3b3325/src/common/pthread_windows.c#L444)):*
<details><summary><i>Previously, ChandKV (Chandra Kumar) wrote…</i></summary>
> That's by design, right? If someone creates a joinable thread and doesn't join we are bound to leak resources right (zombie threads). In the minimum we need to store the thread's exit status somewhere so that we can hand that over to a thread that calls join. Also we need to reserve the index or thread-id, so that a later thread won't take its place.
> 
> If someone is not interested in waiting for a thread, they should create a DETACHED thread. And because our tests are not doing that, yet, our pthread implementation doesn't implement the PTHREAD_CREATE_DETACHED attribute (any attribute for that matter).
> 
</details>
I assume this is resolved, please let me know otherwise.

---

*[src/test/blk_rw_mt/blk_rw_mt.c, line 140 \[r1\]](https://reviewable.io:443/reviews/pmem/nvml/979#-KMTPe86Ptm_VGJsgW_V:-KMolkOeqZGDpdzcKvv7:676635175) ([raw file](https://github.com/pmem/nvml/blob/547e8f63fd769ff4e07ec41afd19244ddb3b3325/src/test/blk_rw_mt/blk_rw_mt.c#L140)):*
<details><summary><i>Previously, ChandKV (Chandra Kumar) wrote…</i></summary>
> Yes, it was lil more code but more explicit. This version silently converts intptr_t to long. I take this as you liked the first version. Thank you, will revert this.
</details>
Done

---

*[src/windows/include/pthread.h, line 159 \[r1\]](https://reviewable.io:443/reviews/pmem/nvml/979#-KMTLxx2rE1onzyAcI0V:-KMolr5w3-WSvKQVXv9-:1735445976) ([raw file](https://github.com/pmem/nvml/blob/547e8f63fd769ff4e07ec41afd19244ddb3b3325/src/windows/include/pthread.h#L159)):*
<details><summary><i>Previously, ChandKV (Chandra Kumar) wrote…</i></summary>
> 
> pthread_t thread;
> pthread_create(&thread, NULL, start_bus, NULL);
> pthread_create(&thread, NULL, start_car, NULL);
> pthread_exit(NULL);
> 
> The above piece of code should successfully call spawn one thread to execute start_bus() and another to execute start_car(). And because the main thread doesn't call pthread_join() we could leak resources but there shouldn't be any uncertainty in what the threads execute.
> 
> If we let the callers allocate memory for pthread_t, the above piece of code could run start_car twice, or crash.
> 
> Also if we had to implement PTHREAD_CREATE_DETACHED tomorrow, above sequence (with PTHREAD_CREATE_DETACHED attribute) should be considered more common.
> 
> Let me know your suggestion.
> 
</details>
Please let me know if you still prefer caller allocating the memory, even if we have the above trade of / issue. The test code that's ported so far doesn't do what's discussed above, so we won't see the issue right away.  Just for record my preference is to **not** let the caller allocate and leave it to the pthread package. For that matter, I was thinking that the caller shouldn't even see the definition of pthread_t but see it as a void *. But this being used internally thought that it's OK.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/979)*
<!-- Sent from Reviewable.io -->

ChandKV commented:
Review status: 11 of 19 files reviewed at latest revision, 12 unresolved discussions.

---

*[src/windows/include/pthread.h, line 159 \[r1\]](https://reviewable.io:443/reviews/pmem/nvml/979#-KMTLxx2rE1onzyAcI0V:-KMowClyg7rk701eJkqf:648297993) ([raw file](https://github.com/pmem/nvml/blob/547e8f63fd769ff4e07ec41afd19244ddb3b3325/src/windows/include/pthread.h#L159)):*
<details><summary><i>Previously, ChandKV (Chandra Kumar) wrote…</i></summary>
> Please let me know if you still prefer caller allocating the memory, even if we have the above trade of / issue. The test code that's ported so far doesn't do what's discussed above, so we won't see the issue right away.  Just for record my preference is to **not** let the caller allocate and leave it to the pthread package. For that matter, I was thinking that the caller shouldn't even see the definition of pthread_t but see it as a void *. But this being used internally thought that it's OK.
</details>
Thinking further, I believe having pthread_t as "void *" for everyone outside pthread library might actually server both the purpose: 1. avoid the above discussed issue 2. Help the readability as discussed in another comment above by @tomaszkapela.

Let me know your thoughts, thank you.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/979)*
<!-- Sent from Reviewable.io -->

krzycz commented:
Reviewed 2 of 10 files at r3.
Review status: 13 of 19 files reviewed at latest revision, 11 unresolved discussions.

---

*[src/common/pthread_windows.c, line 438 \[r3\]](https://reviewable.io:443/reviews/pmem/nvml/979#-KMxsRPWH5QrJa1JQCgd:-KMxsRPXbvyJnxqrSzTQ:1129576429) ([raw file](https://github.com/pmem/nvml/blob/90251b6a927fd2615c53b38f9ea532b9ac4698ab/src/common/pthread_windows.c#L438)):*
> ```C
> 	TlsSetValue(Pthread_self_index, thread_info);
> 
> 	thread_info->start_routine(thread_info->arg);
> ```

Shouldn't it be:
`thread_info->result = thread_info->start_routine(...);` ?

---

*[src/windows/include/pthread.h, line 159 \[r1\]](https://reviewable.io:443/reviews/pmem/nvml/979#-KMTLxx2rE1onzyAcI0V:-KMxKsSOD4s_qPQmAxfp:-1789020618) ([raw file](https://github.com/pmem/nvml/blob/547e8f63fd769ff4e07ec41afd19244ddb3b3325/src/windows/include/pthread.h#L159)):*
<details><summary><i>Previously, ChandKV (Chandra Kumar) wrote…</i></summary>
> Thinking further, I believe having pthread_t as "void *" for everyone outside pthread library might actually server both the purpose: 1. avoid the above discussed issue 2. Help the readability as discussed in another comment above by @tomaszkapela.
> 
> Let me know your thoughts, thank you.
</details>
The example you provided would work in this particular case, and for this particular implementation, but in general this is a misuse of the API, IMHO.  I think we should not design our implementation to support that.  So, I would prefer the caller to allocate memory for *pthread_t*, but - on the other hand - other pthread's implementations for Windows (i.e. Cygwin) do allocations in *pthread_create*, so it's not so bad. Also, we can't forget about other interfaces that we may want implement in the future, like *pthread_self*, *pthread_equal*, etc., which might be easier to implement if pthread_t is just a pointer.

So, I'm fine with your implementation, except one more comment I added on *thread->result*.

---

*[src/windows/include/pthread.h, line 159 \[r3\]](https://reviewable.io:443/reviews/pmem/nvml/979#-KMxsYckWRsUL2NgIxk9:-KMxsYckWRsUL2NgIxkA:1518891258) ([raw file](https://github.com/pmem/nvml/blob/90251b6a927fd2615c53b38f9ea532b9ac4698ab/src/windows/include/pthread.h#L159)):*
> ```objc
> 	void *arg;
> 	void *(*start_routine)(void *);
> 	void *result;
> ```

Is "result" set anywhere?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/979)*
<!-- Sent from Reviewable.io -->

tomaszkapela commented:
Review status: 13 of 19 files reviewed at latest revision, 9 unresolved discussions.

---

*[src/windows/include/pthread.h, line 159 \[r1\]](https://reviewable.io:443/reviews/pmem/nvml/979#-KMTLxx2rE1onzyAcI0V:-KMxyKdnMzAtwNY8qozi:1961701010) ([raw file](https://github.com/pmem/nvml/blob/547e8f63fd769ff4e07ec41afd19244ddb3b3325/src/windows/include/pthread.h#L159)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary>
> The example you provided would work in this particular case, and for this particular implementation, but in general this is a misuse of the API, IMHO.  I think we should not design our implementation to support that.  So, I would prefer the caller to allocate memory for *pthread_t*, but - on the other hand - other pthread's implementations for Windows (i.e. Cygwin) do allocations in *pthread_create*, so it's not so bad. Also, we can't forget about other interfaces that we may want implement in the future, like *pthread_self*, *pthread_equal*, etc., which might be easier to implement if pthread_t is just a pointer.
> 
> So, I'm fine with your implementation, except one more comment I added on *thread->result*.
</details>
After looking into other implementations of pthreads, I'm also starting to like this implementation a bit more ;)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/979)*
<!-- Sent from Reviewable.io -->

ChandKV commented:
Review status: 13 of 19 files reviewed at latest revision, 9 unresolved discussions.

---

*[src/common/pthread_windows.c, line 438 \[r3\]](https://reviewable.io:443/reviews/pmem/nvml/979#-KMxsRPWH5QrJa1JQCgd:-KMz2EM6xZ0XYdxSHBad:1983519746) ([raw file](https://github.com/pmem/nvml/blob/90251b6a927fd2615c53b38f9ea532b9ac4698ab/src/common/pthread_windows.c#L438)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary>
> Shouldn't it be:
> `thread_info->result = thread_info->start_routine(...);` ?
</details>
ah.. my bad. yes it should be that way. Thank you.

---

*[src/windows/include/pthread.h, line 159 \[r1\]](https://reviewable.io:443/reviews/pmem/nvml/979#-KMTLxx2rE1onzyAcI0V:-KMz2N50aobap4WE2Y45:-556193206) ([raw file](https://github.com/pmem/nvml/blob/547e8f63fd769ff4e07ec41afd19244ddb3b3325/src/windows/include/pthread.h#L159)):*
<details><summary><i>Previously, tomaszkapela wrote…</i></summary>
> After looking into other implementations of pthreads, I'm also starting to like this implementation a bit more ;)
</details>
Thank you.

---

*[src/windows/include/pthread.h, line 159 \[r3\]](https://reviewable.io:443/reviews/pmem/nvml/979#-KMxsYckWRsUL2NgIxk9:-KMz3UetBZqgnzqC2G9f:-245196563) ([raw file](https://github.com/pmem/nvml/blob/90251b6a927fd2615c53b38f9ea532b9ac4698ab/src/windows/include/pthread.h#L159)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary>
> Is "result" set anywhere?
</details>
Overlooked it, by bad. Fixed it as part of the above comment. Thank you.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/979)*
<!-- Sent from Reviewable.io -->

krzycz commented:
Reviewed 5 of 10 files at r3, 1 of 1 files at r4.
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/979)*
<!-- Sent from Reviewable.io -->

krzycz commented:
Merged by hand.
